File: 452-findMinArrowShots/findMinArrowShots.js
"use strict";
const findMinArrowShots = (points) => {
    const sorted = points.sort((left, right) => left[0] === right[0] ? left[1] - right[1] : left[0] - right[0]);
    return countArrows(sorted[0][1], sorted);
};
const countArrows = (lastPoint, points) => points.length === 0
    ? 1
    : points[0][0] > lastPoint
        ? 1 + countArrows(points[0][1], points.slice(1))
        : countArrows(min(lastPoint, points[0][1]), points.slice(1));
const min = (a, b) => a < b ? a : b;
// This solution doesn't pass Leetcode due to memory allocation, as each new slice creates a new array :(
// We can go through by using a ptr to the non-visited element.
const findMinArrowShots_with_ptr = (points) => {
    const sorted = points.sort((left, right) => left[0] === right[0] ? left[1] - right[1] : left[0] - right[0]);
    return countArrows_with_ptr(sorted[0][1], sorted, 1);
};
const countArrows_with_ptr = (lastPoint, points, idx) => {
    return idx === points.length
        ? 1
        : points[idx][0] > lastPoint
            ? 1 + countArrows_with_ptr(points[idx][1], points, idx + 1)
            : countArrows_with_ptr(min(lastPoint, points[idx][1]), points, idx + 1);
};
console.log(findMinArrowShots_with_ptr([[10, 16], [2, 8], [1, 6], [7, 12]]), " =? 2");
console.log(findMinArrowShots_with_ptr([[1, 2], [3, 4], [5, 6], [7, 8]]), " =? 4");
console.log(findMinArrowShots_with_ptr([[1, 2], [2, 3], [3, 4], [4, 5]]), " =? 2");
console.log(findMinArrowShots_with_ptr([[9, 12], [1, 10], [4, 11], [8, 12], [3, 9], [6, 9], [6, 7]]), " =? 2");

File: 452-findMinArrowShots/find_min_arrow_shots.rs
fn find_min_arrow_shots(mut points: Vec<Vec<i32>>) -> i32 {
    points.sort();
    count_arrows(points[0][1], points.clone())
}

fn count_arrows(last_point: i32, mut points: Vec<Vec<i32>>) -> i32 {
    println!("{}", last_point);
    match &mut points[..] {
        [] => 1,
        [x, xs @ ..] => {
            println!("{:?} {:?}", x, xs);
            if x[0] > last_point {
                1 + count_arrows(x[1], xs.to_vec())
            } else {
                count_arrows(std::cmp::min(x[1], last_point), xs.to_vec())
            }
        }
    }
}

// It gives tl. It doesn't pass my tests. I don't know why xddddd wtf i hate rust xd

/* I saw this solution, which is beautiful!
pub fn find_min_arrow_shots(mut points: Vec<Vec<i32>>) -> i32 {
    points.sort_unstable_by_key(|b| b[1]);
    points.iter().skip(1).fold((1, points[0][1]), |(rez, right), b|
        if right < b[0] {
            (rez + 1, b[1])
        } else {
            (rez, right)
        }
    ).0
}
*/

#[cfg(test)]
mod tests {
    use crate::find_min_arrow_shots;
    macro_rules! vec_of_vectors {
        ($($x:expr),*) => (vec![$($x.to_vec()), *]);
    }
    #[test]
    fn ex1() {
        assert_eq!(
            find_min_arrow_shots(vec_of_vectors![[10, 16], [2, 8], [1, 6], [7, 12]]),
            2
        )
    }
    #[test]
    fn ex2() {
        assert_eq!(
            !find_min_arrow_shots(vec_of_vectors![[1, 2], [3, 4], [5, 6], [7, 8]]),
            4
        )
    }
    #[test]
    fn ex3() {
        assert_eq!(
            !find_min_arrow_shots(vec_of_vectors![[1, 2], [2, 3], [3, 4], [4, 5]]),
            2
        )
    }
    #[test]
    fn ex4() {
        assert_eq!(
            !find_min_arrow_shots(vec_of_vectors![
                [9, 12],
                [1, 10],
                [4, 11],
                [8, 12],
                [3, 9],
                [6, 9],
                [6, 7]
            ]),
            2
        )
    }
}
